### DISCUSSION TITLE
Understanding how useActor getSnapshot works with entry actions

### DISCUSSION THREAD

**Discussion by dwxw** [statelyai/xstate#4421] [Q&A]
Hello.  I'm seeing some behaviour with Xstate v4 I don't understand.  I've tried to make a simplified example here to demonstrate the problem I have.  A state machine first calls a promise to get an account id as part of `onDone` it assigns the account id to the context.  The next state has an entry action that triggers the rendering of a new React component.  In this component I use useActor to read the account id from the context.  This value will always be wrong on the first render.  Subsequent renders, triggered by the useActor hook have the updated value. If I check the context for the entry action, it has the new context.

```
{
  fetchAccountId: {
    invoke: {
      src: 'fetchAccountId',
      onDone: {
        actions: assign({accountId: (context, event) => event.data.accountId}),
        target: 'displayAccountId'
      }
    }
  },
  displayAccountId: {
    entry: (context) => showAccountIdScreen()
  }
}
```

In my React component
```
const [state] = useActor(waqfService, (actor) => {
    const snapshot = actor.getSnapshot();
    console.log('useActor', snapshot.context.accountId);  // this value is old
    return snapshot;
  });
```

So I'm confused.  Having read the docs, I was under the impression that the context would be updated before transitioning to the next state.  The entry action for the next state does indeed have the new context.  However the following call to useActor triggered by the action has the old value on the first call, before it then having the latest on subsequent renders.  I'm happy the component isn't being rendered until after the entry action is called.

I can fix the problem by creating an intermediate state and putting a delay before transitioning again, but I'd rather not do that.

```
{
  fetchAccountId: {
    invoke: {
      src: 'fetchAccountId',
      onDone: {
        actions: assign({accountId: (context, event) => event.data.accountId}),
        target: 'fixContext'
      }
    }
  },
  fixContext: {    
    after: {
      1: {target: 'displayAccountId'}  
    },
  },
  displayAccountId: {
    entry: (context) => showAccountIdScreen()
  }
}
```

Is this behaviour expected, or any suggestions as a better way to make this work?
Thanks.

**Comment by Andarist** [comment]
It would be best if you could wrap this into a runnable repro case. It's hard to debug the code from just a few code snippets.

**Reply by dwxw** [reply]
I'm struggling to get a full example to show the problem I have.  I'm using React Native with React Navigation, and it's the navigating to the next screen which triggers the render where I have stale data.  I have a simplified not quite working example [here](https://codesandbox.io/s/quizzical-browser-n3t4rl?file=/src/App.js) which hopefully shows better what I'm doing.  

Account Id is set initially when the state machine starts.  Then later a promise sets the value to something else.  On the entry to the final state, I navigate to a new screen which renders a component that reads the account id from the useActor.  However the AccountId component will render the first time will the original id value.  Then immediately render again with the value I expect.

```
TIME                              CONTEXT       REACT
fetch id                           "id1"          -
  |
assign id                          "id2"          -
  |
transition to next state           "id2"          -
  |
entry navigateToScreen             "id2"          -
  | 
render screen                      "id2"         "id1"
  |
render screen                      "id2"         "id2"
```

I don't think I should ever get "id1" in the screen component because that value was overridden in the previous state.  The component that reads the account id is only created from the entry state onwards. Does this behaviour make sense?  I'm worried I shouldn't use entry actions the way I am.
